keyspace/key_space.go

```
@@ -255,20 +269,36 @@ func (ks *KeySpace) WaitForRevision(ctx context.Context, revision int64) error {
 // fixtures; most clients should instead use Watch. Clients must ensure
 // concurrent calls to Apply are not made.
 func (ks *KeySpace) Apply(responses ...clientv3.WatchResponse) error {
-	var hdr = ks.Header
 	var wr clientv3.WatchResponse
 
+	// This will become the new Keyspace.Header after we're done applying all the responses.
+	// We'll validate each response header to make sure that it's consistent with this one,
+	// but we'll only update this header based on either a progress notification or the `ModRevision`
+	// of the events within the response.
+	var lastHeader = ks.Header
+
 	for _, wr = range responses {
-		// Do not apply progress notify events to |hdr| as they may contain
-		// a revision increase without a corresponding modification of this
-		// KeySpace, and we track only modifying revisions.
-		if !wr.IsProgressNotify() {
-			if err := patchHeader(&hdr, wr.Header, false); err != nil {
-				return err
-			}
+		// Sanity check that the response header is consistent, without updating lastHeader.
+		if err := checkHeader(&lastHeader, wr.Header, true); err != nil {
+			return err
+		}
+		if len(wr.Events) == 0 {
```

nit: this cannot ever happen. Just let it panic below instead if it were to.